custom-targets/vertex-ai/model-deployer/deploy.go

```diff
@@ -0,0 +1,304 @@
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("error reading manifest file: %v", err)
+	}
+
+	deployModelRequest := &aiplatform.GoogleCloudAiplatformV1DeployModelRequest{}
+
+	if err = yaml.Unmarshal(data, deployModelRequest); err != nil {
+		return nil, fmt.Errorf("unable to parse deploy model deployModelRequest from manifest file: %v", err)
+	}
+
+	return deployModelRequest, nil
+}
+
+func (d *deployer) applyModel(ctx context.Context, localManifest string) ([]byte, error) {
+
+	deployModelRequest, err := deployModelFromManifest(localManifest)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointRegion, err := fetchRegionFromEndpoint(d.params.endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse region from endpoint resource name: %v", err)
+	}
+
+	aiplatformService, err := newService(ctx, endpointRegion)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create service: %v", err)
+	}
+
+	if d.req.Percentage != 100 {
+		previousModel, err := determinePreviousModel(aiplatformService, d.params.endpoint, deployModelRequest.DeployedModel.Model)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get previous model to canary against: %v", err)
+		}
+		previousPercentage, ok := deployModelRequest.TrafficSplit["previous-model"]
+		if !ok {
+			return nil, fmt.Errorf("expected input manifest trafficSplit stanza to have a 'previous-model' entry but did not find it")
+		}
+		delete(deployModelRequest.TrafficSplit, "previous-model")
+		deployModelRequest.TrafficSplit[previousModel] = previousPercentage
+	}
+	op, err := aiplatformService.Projects.Locations.Endpoints.DeployModel(d.params.endpoint, deployModelRequest).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to deploy model: %v", err)
+	}
+
+	if err := Poll(ctx, aiplatformService, op); err != nil {
+		return nil, err
+	}
+
+	endpoint, err := aiplatformService.Projects.Locations.Endpoints.Get(d.params.endpoint).Do()
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch endpoint: %v", err)
+	}
+
+	var modelsToUndeploy = map[string]bool{}
+	for _, dm := range endpoint.DeployedModels {
+		modelsToUndeploy[dm.Id] = true
+	}
+
+	for id, _ := range endpoint.TrafficSplit {
+		delete(modelsToUndeploy, id)
+	}
+
+	undeployedCount := 0
+	err = nil
+	var lros []*aiplatform.GoogleLongrunningOperation
+	for id, _ := range modelsToUndeploy {
+
+		undeployRequest := &aiplatform.GoogleCloudAiplatformV1UndeployModelRequest{DeployedModelId: id}
+		lro, lroErr := aiplatformService.Projects.Locations.Endpoints.UndeployModel(d.params.endpoint, undeployRequest).Do()
+		if err != nil {
+			fmt.Printf("error undeploying model: %v\n", err)
+			err = lroErr
+			undeployedCount += 1
+		} else {
+			lros = append(lros, lro)
+		}
+	}
+
+	for pollErr := range pollChan(ctx, aiplatformService, lros...) {
+		if pollErr != nil {
+			fmt.Printf("Error in undeploy model operation: %v", err)
+			err = pollErr
+		}
+	}
```

Comment:
Similar to comment above, you can have a single undeploy method defined in the AI service file that handles all this so it doesn't detract from the rest of the deploy logic.